### ISSUE TITLE
[BUG} Auto_id feature does not work correctly when loading deserialized object

### ISSUE THREAD

**Issue by PenguinWithATie** [clementwanjau/tree-ds#11]
**Describe the bug**
Auto_id feature does not work correctly when loading deserialized object, it will generate duplicate ids

**To Reproduce**
1. serialize a tree made with auto_id nodes
2. save to a file
3. In a different process, open the tree and add a node
4.  node will have duplicated id

**Expected behavior**
node should be added correctly

Edit: added missing step

**Comment by clementwanjau** [comment]
Noted. This will be fixed in the next release.

**Comment by clementwanjau** [comment]
This must be some edge case because writing a test to simulate the reproduction of the stated bug shows that the id is unique for some reason.

The test:

```rust
    #[cfg(all(feature = "auto_id", feature = "serde"))]
    #[test]
    fn test_deserialize_tree_with_auto_id() {
        let mut tree = Tree::<i32, i32>::new(Some("Sample Tree"));
        let root = tree
            .add_node(Node::new_with_auto_id(Some(2)), None)
            .unwrap();
        let child_1 = tree
            .add_node(Node::new_with_auto_id(Some(3)), Some(&root))
            .unwrap();
        let child_2 = tree
            .add_node(Node::new_with_auto_id(Some(4)), Some(&child_1))
            .unwrap();
        let child_3 = tree
            .add_node(Node::new_with_auto_id(Some(5)), Some(&child_2))
            .unwrap();
        let serialized_tree = serde_json::to_string(&tree).unwrap();
        println!("*****************************************");
        println!("SERIALIZED TREE");
        println!("-----------------------------------------");
        println!("{}", tree);
        let mut deserialized_tree: Tree<i32, i32> = serde_json::from_str(&serialized_tree).unwrap();
        println!("*****************************************");
        println!("DESERIALIZED TREE");
        println!("-----------------------------------------");
        println!("{}", deserialized_tree.to_string());
        deserialized_tree
            .add_node(Node::new_with_auto_id(Some(6)), Some(&child_3))
            .unwrap();
        println!("*****************************************");
        println!("DESERIALIZED TREE AFTER ADDING NEW NODE");
        println!("-----------------------------------------");
        println!("{}", deserialized_tree);
    }

```

The result:
```text
*****************************************
SERIALIZED TREE
-----------------------------------------
Sample Tree
***********
1: 2
└── 2: 3
    └── 3: 4
        └── 4: 5

*****************************************
DESERIALIZED TREE
-----------------------------------------
Sample Tree
***********
1: 2
└── 2: 3
    └── 3: 4
        └── 4: 5

*****************************************
DESERIALIZED TREE AFTER ADDING NEW NODE
-----------------------------------------
Sample Tree
***********
1: 2
└── 2: 3
    └── 3: 4
        └── 4: 5
            └── 5: 6

```

Additional information may be useful to reproduce the bug. For instance:
- Do you store the serialized tree to disk or db and then read it back?

**Comment by clementwanjau** [comment]
Additional information may be useful to reproduce the bug. For instance:
- Do you store the serialized tree to disk or db and then read it back?

**Comment by PenguinWithATie** [comment]
yes it must be serialized to disc then read (so new is not called) then it will reproduce

**Comment by clementwanjau** [comment]
This will be fixed in the next released.

**Comment by clementwanjau** [comment]
This bug has been fixed in `v0.1.5`. The `auto_id` feature now guarantees uniqueness in id generation. 

The usage has also changed a bit. Now trees with automated node ids are of type `Tree::<AutomatedId, T>` where T is the type of the node value. 

You can alias the tree type in your code to write cleaner code:

```rust
use tree_ds::prelude::*;

type AutoTree<T> = Tree<AutomatedId, T>;

let mut tree = AutoTree::<&str>::new(Some("Sample Tree"));
tree.add_node(Node::new_with_auto_id(Some("Node Value")))?;
```
